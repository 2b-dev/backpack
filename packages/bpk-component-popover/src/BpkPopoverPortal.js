import focusStore from 'a11y-focus-store';
import focusScope from 'a11y-focus-scope';
import { Portal } from 'bpk-react-utils';
import React, { PropTypes, Component } from 'react';
import Tether, { getArrowPositionCallback } from 'bpk-tether';

import './bpk-popover.scss';
import BpkPopover from './BpkPopover';
import { ARROW_ID } from './constants';

class BpkPopoverPortal extends Component {
  constructor() {
    super();

    this.tether = null;

    this.onOpen = this.onOpen.bind(this);
    this.beforeClose = this.beforeClose.bind(this);
  }

  onOpen(popoverElement, targetElement) {
    this.tether = new Tether({
      classPrefix: 'bpk-popover-tether',
      element: popoverElement,
      target: targetElement,
      ...this.props.tetherOptions,
    });

    this.tether.on('position', getArrowPositionCallback(popoverElement, ARROW_ID, 'bpk-popover-tether'));

    this.tether.position();
    // setTimeout(() => {
    //   this.tether.position();
    // });

    focusStore.storeFocus();
    focusScope.scopeFocus(popoverElement);
  }

  beforeClose(done) {
    this.tether.destroy();
    this.tether = null;

    focusScope.unscopeFocus();
    focusStore.restoreFocus();

    done();
  }

  render() {
    const { target, isOpen, onClose, ...rest } = this.props;

    delete rest.onClose;
    delete rest.tetherOptions;

    return (
      <Portal
        target={target}
        isOpen={isOpen}
        onClose={onClose}
        onOpen={this.onOpen}
        beforeClose={this.beforeClose}
      >
        <BpkPopover onClose={onClose} {...rest} />
      </Portal>
    );
  }
}

BpkPopoverPortal.propTypes = {
  target: PropTypes.element.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  tetherOptions: PropTypes.shape({
    attachment: PropTypes.string,
    targetAttachment: PropTypes.string,
    offset: PropTypes.string,
    constraints: PropTypes.array,
  }),
};

BpkPopoverPortal.defaultProps = {
  tetherOptions: {
    attachment: 'top center',
    constraints: [
      {
        to: 'window',
        attachment: 'together',
        pin: true,
      },
    ],
  },
};

export default BpkPopoverPortal;
