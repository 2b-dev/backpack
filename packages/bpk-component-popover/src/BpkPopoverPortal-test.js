import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

jest.mock('tether');
jest.mock('a11y-focus-store', () => ({
  storeFocus: jest.fn(),
  restoreFocus: jest.fn(),
}));
jest.mock('a11y-focus-scope', () => ({
  scopeFocus: jest.fn(),
  unscopeFocus: jest.fn(),
}));

/* eslint-disable import/first */
import BpkPopoverPortal from './BpkPopoverPortal';
/* eslint-enable */

describe('BpkPopoverPortal', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <BpkPopoverPortal
        id="my-popover"
        target={<div>target</div>}
        isOpen={false}
        onClose={() => null}
        closeButtonText="Close"
      >
        <div>My popover content</div>
      </BpkPopoverPortal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should trap and restore focus', (done) => {
    const focusStore = require('a11y-focus-store'); // eslint-disable-line global-require
    const focusScope = require('a11y-focus-scope'); // eslint-disable-line global-require

    const portal = mount(
      <BpkPopoverPortal
        id="my-popover"
        target={<div>target</div>}
        isOpen={false}
        onClose={() => null}
        closeButtonText="Close"
      >
        <div>My popover content</div>
      </BpkPopoverPortal>,
    );

    expect(portal.instance().tether).toBeNull();
    expect(focusStore.storeFocus).not.toHaveBeenCalled();
    expect(focusScope.scopeFocus).not.toHaveBeenCalled();
    expect(focusStore.restoreFocus).not.toHaveBeenCalled();
    expect(focusScope.unscopeFocus).not.toHaveBeenCalled();

    portal.setProps({ isOpen: true }, () => {
      expect(focusStore.storeFocus).toHaveBeenCalled();
      expect(focusScope.scopeFocus).toHaveBeenCalled();

      portal.setProps({ isOpen: false }, () => {
        expect(focusStore.restoreFocus).toHaveBeenCalled();
        expect(focusScope.unscopeFocus).toHaveBeenCalled();
        done();
      });
    });
  });
});
