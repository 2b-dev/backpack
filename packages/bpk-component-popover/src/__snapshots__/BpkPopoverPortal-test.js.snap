exports[`BpkPopoverPortal should render correctly 1`] = `
<div>
  target
</div>
`;
